### DISCUSSION TITLE
genes reported in noUniprot can actually be found in data/uniprot.tsv

### DISCUSSION THREAD

**Discussion by PkiwiBird** [SysBioChalmers/GECKO#377] [Q&A] [answered]
When running:
`[ecModel, noUniprot] = makeEcModel(model, false);`
there is this warning:
```
WARNING: The 30 gene(s) reported in noUniprot cannot be found in data/uniprot.tsv, these will
not be enzyme-constrained. If you intend to use different Uniprot data (e.g. from a
different proteome, make sure you first delete the existing data/uniprot.tsv file. 
```
1. Some of the missing IDs in `noUniprot` exist in` uniprot.tsv `file, but are not detected as such maybe because there are several equivalent geneIDs split by ';', e.g.
`{'100137049'}` exists in row `P0C869	100137049;8681	3.1.1.4; 3.1.1.5	87978	MAVAEV....`
`{'100526794'}` exists in row `Q96P26	100526794;93034	3.1.3.5	68804	MSQTSL...`
How to solve this? Is it enough to split into two rows for each GeneID (2nd column) or would that create problems because the Entry (1st column) would be repeated?

2. Also, there is one gene `{'6818'     }` that it shows in two rows:
```
P0DMM9	445329;6818	2.8.2.1	34196	MELIQ...
P0DMN0	445329;6818	2.8.2.1	34196	MELIQ...
```
Can this create problems?

3. There are Gene ids that are not in uniprot.tsv file. e.g. `{'100288072'}`
How to deal with those?

Important parameters in ecModelGEMAdapter.m:
```
obj.params.complex.taxonomicID = 9606;
obj.params.kegg.ID = 'hsa';
obj.params.kegg.geneID = 'kegg';
obj.params.uniprot.type = 'proteome';
obj.params.uniprot.ID = 'UP000005640';
obj.params.uniprot.geneIDfield = 'xref_geneid';
```

**Comment by edkerk** [comment, marked as answer]
1. Duplicate Uniprot IDs are not allowed. You can maybe first check if both of the genes (`100137049` and `8681`) are included in the model. If e.g. `8681` is not present, then remove it from the `uniprot.tsv`. If both genes are present, then a work-around would be to change the gene association in the conventional GEM, e.g. replace all instances of `8681` with `100137049`, and then again remove `8681` from the `uniprot.tsv`.
2. The first problem is of course that there are two genes (`445329` and `6818`) assigned to the same Uniprot ID. If you would remove `445329` from both of the entries, then GECKO would automatically choose the first protein it finds: `P0DMM9`.
3. In order to apply ec-constraints in relation to that protein, you should manually add the necessary information to the `uniprot.tsv` file. You might not raise a representative example though, as `100288072` seems to be in [Uniprot](https://www.uniprot.org/uniprotkb/A0A8I5KQB5/entry) and should be included with the `UP000005640` proteome? Did you maybe have `obj.params.uniprot.reviewed = true;`? Try setting it to `false`.

**Reply by PkiwiBird** [reply]
3. Even after setting obj.params.uniprot.reviewed = false that `100288072` is in the `noUniprot` and doesn't exist in the `uniprot.tsv`

**Reply by edkerk** [reply]
No idea why it's not being picked up. Easiest solution might be to add the few missing proteins to `uniprot.tsv` yourself.
